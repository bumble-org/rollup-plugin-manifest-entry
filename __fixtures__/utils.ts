import { resolve } from 'path'
import { OutputAsset, OutputChunk, OutputBundle } from 'rollup'

export const getExtPath = (path: string): string =>
  resolve(__dirname, 'extensions', path)

/**  Make relative to project root */
export const getRelative = (p: string): string =>
  p.replace(process.cwd() + '/', '')

export function byFileName(n: string) {
  return ({ fileName }: OutputAsset | OutputChunk): boolean =>
    fileName === n
}

/**
 * Get the source of an OutputAsset as a string
 */
export const getAssetSource = (
  key: string,
  bundle: OutputBundle,
): string => {
  const asset = bundle[key] as OutputAsset

  if (!asset) {
    throw new Error(`Unable to find ${key} in bundle`)
  }

  if (asset.source instanceof Buffer) {
    return asset.source.toString('utf8')
  } else {
    return asset.source
  }
}
