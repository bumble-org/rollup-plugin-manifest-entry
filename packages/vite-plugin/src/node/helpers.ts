import { simple } from 'acorn-walk'
import { createHash as _hash } from 'crypto'
import debug from 'debug'
import fg from 'fast-glob'
import { AcornNode, PluginContext } from 'rollup'
import v8 from 'v8'
import type {
  ManifestV3,
  WebAccessibleResourceById,
  WebAccessibleResourceByMatch,
} from './manifest'
import type {
  AcornIdentifier,
  AcornLiteral,
  AcornMemberExpression,
  AcornTemplateElement,
  ManifestFiles,
} from './types'

export const _debug = (id: string) => debug('crx').extend(id)

export const structuredClone = <T>(obj: T): T => {
  return v8.deserialize(v8.serialize(obj))
}

export const createHash = (data: string, length = 5): string =>
  _hash('sha1')
    .update(data)
    .digest('base64')
    .replace(/[^A-Za-z0-9]/g, '')
    .slice(0, length)

export const isString = (x: unknown): x is string => typeof x === 'string'

type Falsy = false | 0 | '' | null | undefined
export const isTruthy = <T>(x: T | Falsy): x is T => !!x

export const isPresent = <T>(x: T | null | undefined): x is T =>
  x !== null && typeof x !== 'undefined'

export function isObject<T>(
  value: T,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
): value is Extract<T, Record<string, any>> {
  return Object.prototype.toString.call(value) === '[object Object]'
}

export const isResourceByMatch = (
  x: WebAccessibleResourceById | WebAccessibleResourceByMatch,
): x is WebAccessibleResourceByMatch => 'matches' in x

export async function manifestFiles(
  manifest: ManifestV3,
  options: fg.Options = {},
): Promise<ManifestFiles> {
  // JSON
  let locales: string[] = []
  if (manifest.default_locale)
    locales = await fg('_locales/**/messages.json', options)

  const rulesets =
    manifest.declarative_net_request?.rule_resources.flatMap(
      ({ path }) => path,
    ) ?? []

  const contentScripts = manifest.content_scripts?.flatMap(({ js }) => js) ?? []
  const contentStyles = manifest.content_scripts?.flatMap(({ css }) => css)
  const serviceWorker = manifest.background?.service_worker
  const htmlPages = htmlFiles(manifest)

  const icons = [
    Object.values(isString(manifest.icons) ? [manifest.icons] : manifest.icons ?? {}) as string[],
    Object.values(isString(manifest.action?.default_icon) ? [manifest.action?.default_icon] : manifest.action?.default_icon ?? {}) as string[],
  ].flat()

  let webAccessibleResources: string[] = []
  if (manifest.web_accessible_resources) {
    const resources = await Promise.all(
      manifest.web_accessible_resources
        .flatMap(({ resources }) => resources!)
        .map(async (r) => {
          // don't copy node_modules, etc
          if (['*', '**/*'].includes(r)) return undefined
          if (fg.isDynamicPattern(r)) return fg(r, options)
          return r
        }),
    )
    webAccessibleResources = resources.flat().filter(isString)
  }

  return {
    contentScripts: [...new Set(contentScripts)].filter(isString),
    contentStyles: [...new Set(contentStyles)].filter(isString),
    html: htmlPages,
    icons: [...new Set(icons)].filter(isString),
    locales: [...new Set(locales)].filter(isString),
    rulesets: [...new Set(rulesets)].filter(isString),
    background: [serviceWorker].filter(isString),
    webAccessibleResources,
  }
}

export function htmlFiles(manifest: ManifestV3): string[] {
  const files = [
    manifest.action?.default_popup,
    Object.values(manifest.chrome_url_overrides ?? {}),
    manifest.devtools_page,
    manifest.options_page,
    manifest.options_ui?.page,
    manifest.sandbox?.pages,
  ]
    .flat()
    .filter(isString)
    .map((s) => s.split('#')[0])
    .sort()
  return [...new Set(files)]
}

export function isMemberExpression(n: AcornNode): n is AcornMemberExpression {
  return n.type === 'MemberExpression'
}

export function isLiteral(n: AcornNode): n is AcornLiteral {
  return n.type === 'Literal'
}

export function isIdentifier(n: AcornNode): n is AcornIdentifier {
  return n.type === 'Identifier'
}

export function decodeManifest(this: PluginContext, code: string): ManifestV3 {
  const tree = this.parse(code)

  let literal: AcornLiteral | undefined
  let templateElement: AcornTemplateElement | undefined

  simple(tree, {
    Literal(node) {
      literal = node as AcornLiteral
    },
    TemplateElement(node) {
      templateElement = node as AcornTemplateElement
    },
  })

  let manifestJson: string | undefined = literal?.value
  if (!manifestJson) manifestJson = templateElement?.value?.cooked
  if (!manifestJson) throw new Error('unable to parse manifest code')

  let result: ManifestV3 = JSON.parse(manifestJson)
  if (typeof result === 'string') result = JSON.parse(result)
  return result
}

export function encodeManifest(manifest: ManifestV3): string {
  const json = JSON.stringify(JSON.stringify(manifest))
  return `export default ${json}`
}

/**
 * [Strip paths for `web_accessible_resources`'s `matches` Â· Issue
 * #282](https://github.com/crxjs/chrome-extension-tools/issues/282)
 */
export const stubMatchPattern = (pattern: string): string => {
  /**
   * Allow <all_urls> in matches section. 
   * [Issue #459](https://github.com/crxjs/chrome-extension-tools/issues/459)
   */
  if (pattern === "<all_urls>") {
    return pattern;
  }
  const [schema, rest] = pattern.split('://')
  const [origin, pathname] = rest.split('/')
  const root = `${schema}://${origin}`
  return pathname ? `${root}/*` : root
}
