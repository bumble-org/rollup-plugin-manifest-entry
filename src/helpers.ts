import { OutputAsset, OutputChunk, OutputBundle } from 'rollup'
import prettier from 'prettier'
import { CheerioFile } from './html-inputs'
import { ChromeExtensionManifest } from './manifest'

export const not = <T>(fn: (x: T) => boolean) => (x: T) => !fn(x)

export function isChunk(
  x: OutputChunk | OutputAsset,
): x is OutputChunk {
  return x.type === 'chunk'
}

export function isAsset(
  x: OutputChunk | OutputAsset,
): x is OutputChunk {
  return x.type === 'asset'
}

export function isString(x: any): x is string {
  return typeof x === 'string'
}

export function isJsonFilePath(x: any): x is string {
  return isString(x) && x.endsWith('json')
}

export const formatHtml = ($: CheerioFile) =>
  prettier.format($.html(), { parser: 'html' })

/**
 * Update the manifest source in the output bundle
 */
export const updateManifest = (
  updater: (
    manifest: ChromeExtensionManifest,
  ) => ChromeExtensionManifest,
  bundle: OutputBundle,
  handleError?: (message: string) => void,
): OutputBundle => {
  try {
    const manifestKey = 'manifest.json'
    const manifestAsset = bundle[manifestKey] as OutputAsset

    if (!manifestAsset) {
      throw new Error(
        'No manifest.json in the rollup output bundle.',
      )
    }

    const manifest = JSON.parse(
      manifestAsset.source as string,
    ) as ChromeExtensionManifest

    const result = updater(manifest)

    manifestAsset.source = JSON.stringify(result, undefined, 2)
  } catch (error) {
    if (handleError) {
      handleError(error.message)
    } else {
      throw error
    }
  }

  return bundle
}
