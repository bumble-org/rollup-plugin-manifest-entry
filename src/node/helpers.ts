import { simple } from 'acorn-walk'
import { createHash as _hash } from 'crypto'
import debug from 'debug'
import fg from 'fast-glob'
import { PluginContext } from 'rollup'
import v8 from 'v8'
import type {
  ManifestV3,
  WebAccessibleResourceById,
  WebAccessibleResourceByMatch,
} from './manifest'
import type { AcornLiteral, ManifestFiles } from './types'

export const _debug = (id: string) => debug('crx').extend(id)

export const structuredClone = <T>(obj: T): T => {
  return v8.deserialize(v8.serialize(obj))
}

export const createHash = (data: string, length = 5): string =>
  _hash('sha1')
    .update(data)
    .digest('base64')
    .replace(/[^A-Za-z0-9]/g, '')
    .slice(0, length)

export const isString = (x: unknown): x is string => typeof x === 'string'

type Falsy = false | 0 | '' | null | undefined
export const isTruthy = <T>(x: T | Falsy): x is T => !!x

export function isObject<T>(
  value: T,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
): value is Extract<T, Record<string, any>> {
  return Object.prototype.toString.call(value) === '[object Object]'
}

export const isResourceByMatch = (
  x: WebAccessibleResourceById | WebAccessibleResourceByMatch,
): x is WebAccessibleResourceByMatch => 'matches' in x

export async function manifestFiles(
  manifest: ManifestV3,
  options: fg.Options = {},
): Promise<ManifestFiles> {
  // JSON
  let locales: string[] = []
  if (manifest.default_locale)
    locales = await fg('_locales/**/messages.json', options)

  const rulesets =
    manifest.declarative_net_request?.rule_resources.flatMap(
      ({ path }) => path,
    ) ?? []

  const contentScripts = manifest.content_scripts?.flatMap(({ js }) => js) ?? []
  const contentStyles = manifest.content_scripts?.flatMap(({ css }) => css)
  const serviceWorker = manifest.background?.service_worker
  const htmlPages = htmlFiles(manifest)

  const icons = [
    Object.values(manifest.icons ?? {}) as string[],
    Object.values(manifest.action?.default_icon ?? {}) as string[],
  ].flat()

  return {
    contentScripts: [...new Set(contentScripts)].filter(isString),
    contentStyles: [...new Set(contentStyles)].filter(isString),
    html: htmlPages,
    icons: [...new Set(icons)].filter(isString),
    locales: [...new Set(locales)].filter(isString),
    rulesets: [...new Set(rulesets)].filter(isString),
    background: [serviceWorker].filter(isString),
  }
}

export function htmlFiles(manifest: ManifestV3): string[] {
  const files = [
    manifest.action?.default_popup,
    Object.values(manifest.chrome_settings_overrides ?? {}),
    manifest.devtools_page,
    manifest.options_page,
    manifest.options_ui?.page,
  ]
    .flat()
    .filter(isString)
    .sort()
  return [...new Set(files)]
}

export function decodeManifest(this: PluginContext, code: string): ManifestV3 {
  const tree = this.parse(code)
  let literal: AcornLiteral | undefined
  simple(tree, {
    Literal(n) {
      literal = n as AcornLiteral
    },
  })
  if (!literal) throw new Error('unable to parse manifest code')
  let result: ManifestV3 = JSON.parse(literal.value)
  if (typeof result === 'string') result = JSON.parse(result)
  return result
}

export function encodeManifest(manifest: ManifestV3): string {
  const json = JSON.stringify(JSON.stringify(manifest))
  return `export default ${json}`
}
