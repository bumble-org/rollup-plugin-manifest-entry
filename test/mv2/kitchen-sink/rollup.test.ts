import { isAsset, isChunk } from '$src/helpers'
import { getRollupOutput } from '$test/helpers/getRollupOutput'
import { jestSetTimeout } from '$test/helpers/timeout'
import { byFileName } from '$test/helpers/utils'
import { OutputAsset } from 'rollup'

jestSetTimeout(30000)

test('bundles chunks and assets', async () => {
  const { output } = await getRollupOutput(
    __dirname,
    'rollup.config.js',
  )

  // Chunks
  const chunks = output.filter(isChunk)
  expect(output.find(byFileName('background.js'))).toBeDefined()
  expect(output.find(byFileName('content.js'))).toBeDefined()
  expect(output.find(byFileName('options1.js'))).toBeDefined()
  expect(output.find(byFileName('options2.js'))).toBeDefined()
  expect(output.find(byFileName('options3.js'))).toBeDefined()
  expect(output.find(byFileName('options4.js'))).toBeDefined()
  expect(output.find(byFileName('popup/popup.js'))).toBeDefined()
  expect(
    output.find(byFileName('devtools/devtools1.js')),
  ).toBeDefined()
  expect(
    output.find(byFileName('devtools/devtools2.js')),
  ).toBeDefined()

  const imported = output.find(({ fileName }) =>
    fileName.includes('imported'),
  )
  // Chunk name should not be double hashed
  expect(imported?.fileName).toMatch(
    // eslint-disable-next-line no-useless-escape
    /^modules\/imported-[a-z0-9\-]+?\.js$/,
  )

  // 9 chunks + one shared import (imported.js)
  expect(chunks.length).toBe(10)

  // Assets
  const assets = output.filter(isAsset)

  expect(output.find(byFileName('manifest.json'))).toBeDefined()

  expect(output.find(byFileName('asset.js'))).toBeDefined()

  expect(
    output.find(byFileName('popup/popup.html')),
  ).toBeDefined()
  expect(
    output.find(byFileName('devtools/devtools.html')),
  ).toBeDefined()

  expect(
    output.find(byFileName('images/icon-main-16.png')),
  ).toBeDefined()
  expect(
    output.find(byFileName('images/icon-main-48.png')),
  ).toBeDefined()
  expect(
    output.find(byFileName('images/icon-main-128.png')),
  ).toBeDefined()
  expect(
    output.find(byFileName('images/favicon.ico')),
  ).toBeDefined()
  expect(
    output.find(byFileName('images/favicon.png')),
  ).toBeDefined()

  expect(output.find(byFileName('options.html'))).toBeDefined()
  expect(output.find(byFileName('options.css'))).toBeDefined()
  expect(output.find(byFileName('content.css'))).toBeDefined()
  expect(output.find(byFileName('options.png'))).toBeDefined()
  expect(output.find(byFileName('options.jpg'))).toBeDefined()

  expect(
    output.find(byFileName('_locales/en/messages.json')),
  ).toBeDefined()
  expect(
    output.find(byFileName('_locales/es/messages.json')),
  ).toBeDefined()

  // assets above + content wrapper script
  expect(assets.length).toBe(18)

  const manifestAsset = output.find(
    byFileName('manifest.json'),
  ) as OutputAsset
  const manifestSource = manifestAsset.source as string
  const manifest: chrome.runtime.Manifest =
    JSON.parse(manifestSource)

  expect(manifest).toMatchObject({
    content_security_policy:
      "script-src 'self'; object-src 'self'",
  })
})
