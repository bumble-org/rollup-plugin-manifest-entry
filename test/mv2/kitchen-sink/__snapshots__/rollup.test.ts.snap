// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`manifest vs output: _locales/en/messages.json 1`] = `"{\\"extName\\":{\\"message\\":\\"MV3 with everything\\",\\"description\\":\\"Chrome extension name\\"},\\"extDescription\\":{\\"message\\":\\"MV3 Chrome Extension with all the options.\\",\\"description\\":\\"Chrome extension description\\"}}"`;

exports[`manifest vs output: _locales/es/messages.json 1`] = `"{\\"extName\\":{\\"message\\":\\"MV3 con todo\\"},\\"extDescription\\":{\\"message\\":\\"MV3 Chrome Extension con todo los opciones\\"}}"`;

exports[`manifest vs output: 00 - manifest 1`] = `
Object {
  "background": Object {
    "scripts": Array [
      "background.esm-wrapper.js",
    ],
  },
  "browser_action": Object {
    "default_icon": Object {
      "16": "images/icon-main-16.png",
    },
    "default_popup": "popup/popup.html",
  },
  "content_scripts": Array [
    Object {
      "js": Array [
        "assets/content.esm-wrapper-a6ac6bbe.js",
      ],
      "matches": Array [
        "https://www.google.com/*",
      ],
    },
    Object {
      "css": Array [
        "content.css",
      ],
      "js": Array [
        "content.js",
      ],
      "matches": Array [
        "https://www.yahoo.com/*",
      ],
    },
  ],
  "content_security_policy": "script-src 'self'; object-src 'self'",
  "default_locale": "en",
  "description": "mv2-kitchen-sink chrome extension",
  "devtools_page": "devtools/devtools.html",
  "icons": Object {
    "128": "images/icon-main-128.png",
    "16": "images/icon-main-16.png",
    "48": "images/icon-main-48.png",
  },
  "manifest_version": 2,
  "name": "mv2-kitchen-sink",
  "options_page": "options.html",
  "permissions": Array [
    "webRequest",
    "webRequestBlocking",
    "storage",
    "contextMenus",
    "bookmarks",
    "cookies",
  ],
  "version": "1.0.0",
  "web_accessible_resources": Array [
    "options.jpg",
    "fonts/*.ttf",
    "fonts/*.otf",
    "content.js",
  ],
}
`;

exports[`manifest vs output: 01 - files from manifest 1`] = `
Object {
  "BACKGROUND": Array [
    "background.esm-wrapper.js",
  ],
  "CONTENT": Array [
    "assets/content.esm-wrapper-a6ac6bbe.js",
    "content.js",
  ],
  "CSS": Array [
    "content.css",
  ],
  "HTML": Array [
    "options.html",
    "devtools/devtools.html",
    "popup/popup.html",
  ],
  "IMAGE": Array [
    "images/icon-main-16.png",
    "images/icon-main-48.png",
    "images/icon-main-128.png",
  ],
  "JSON": Array [
    "_locales/**/messages.json",
  ],
}
`;

exports[`manifest vs output: 02 - files from output 1`] = `
Array [
  "_locales/en/messages.json",
  "_locales/es/messages.json",
  "asset.js",
  "assets/content.esm-wrapper-a6ac6bbe.js",
  "background.esm-wrapper.js",
  "background.js",
  "content.css",
  "content.js",
  "devtools/devtools.html",
  "devtools/devtools1.js",
  "devtools/devtools2.js",
  "images/favicon.ico",
  "images/favicon.png",
  "images/icon-main-128.png",
  "images/icon-main-16.png",
  "images/icon-main-48.png",
  "manifest.json",
  "modules/imported-fc5abad3.js",
  "options.css",
  "options.html",
  "options.jpg",
  "options.png",
  "options1.js",
  "options2.js",
  "options3.js",
  "options4.js",
  "popup/popup.html",
  "popup/popup.js",
]
`;

exports[`manifest vs output: 03 - manifest files match output 1`] = `
Object {
  "assets/content.esm-wrapper-a6ac6bbe.js": true,
  "background.esm-wrapper.js": true,
  "content.css": true,
  "content.js": true,
  "devtools/devtools.html": true,
  "images/icon-main-128.png": true,
  "images/icon-main-16.png": true,
  "images/icon-main-48.png": true,
  "options.html": true,
  "popup/popup.html": true,
}
`;

exports[`manifest vs output: assets/content.esm-wrapper-a6ac6bbe.js 1`] = `
"(function () {
  'use strict';

  const importPath = \\"content.js\\";

  const url = chrome.runtime.getURL(importPath);
  import(
    /* @vite-ignore */
    url
  ).catch((err) => {
    console.warn(\`Could not import \${importPath} from \${chrome.runtime.getManifest().name}\`);
    console.error(err);
  });

})();
"
`;

exports[`manifest vs output: background.esm-wrapper.js 1`] = `
"(function () {
  'use strict';

  function getDeepMatches(object, pred, excludeKeys) {
    const keys = typeof object === \\"object\\" && object ? Object.keys(object) : [];
    return keys.length ? keys.filter((key) => !excludeKeys.includes(key)).reduce((r, key) => {
      const target = object[key];
      if (target && pred(target)) {
        return [...r, target];
      } else {
        return [
          ...r,
          ...getDeepMatches(target, pred, excludeKeys)
        ];
      }
    }, []) : [];
  }

  const importPath = \\"./background.js\\";

  const events = getDeepMatches(chrome, (x) => typeof x === \\"object\\" && \\"addListener\\" in x, [\\"webRequest\\", \\"extension\\"]);
  const triggerEvents = captureEvents(events);
  import(importPath).then(async () => {
    triggerEvents();
  });
  function captureEvents(events2) {
    const captured = events2.map(captureEvent);
    return () => captured.forEach((t) => t());
    function captureEvent(event) {
      let isCapturePhase = true;
      const callbacks = /* @__PURE__ */ new Map();
      const eventArgs = /* @__PURE__ */ new Set();
      event.addListener(handleEvent);
      function handleEvent(...args) {
        if (isCapturePhase) {
          eventArgs.add(args);
          if (typeof args[2] === \\"function\\") {
            return true;
          } else {
            return false;
          }
        } else {
          return callListeners(...args);
        }
      }
      function callListeners(...args) {
        let isAsyncCallback = false;
        callbacks.forEach((options, cb) => {
          try {
            isAsyncCallback = cb(...args) || isAsyncCallback;
          } catch (error) {
            console.error(error);
          }
        });
        if (!isAsyncCallback && typeof args[2] === \\"function\\") {
          args[2]();
        }
        return isAsyncCallback;
      }
      function triggerEvents2() {
        eventArgs.forEach((args) => {
          callListeners(...args);
        });
        isCapturePhase = false;
        eventArgs.clear();
      }
      event.addListener = function addListener(cb, ...options) {
        callbacks.set(cb, options);
      };
      event.hasListeners = function hasListeners() {
        return callbacks.size > 0;
      };
      event.hasListener = function hasListener(cb) {
        return callbacks.has(cb);
      };
      event.removeListener = function removeListener(cb) {
        callbacks.delete(cb);
      };
      event.__isCapturedEvent = true;
      return triggerEvents2;
    }
  }

})();
"
`;

exports[`manifest vs output: background.js 1`] = `
"import { x } from './modules/imported-fc5abad3.js';

console.log(x);
console.log(\\"background.js\\");
chrome.storage.local.clear();
"
`;

exports[`manifest vs output: content.css 1`] = `
"body {
    color: blueviolet;
  }
  "
`;

exports[`manifest vs output: content.js 1`] = `
"import { x } from './modules/imported-fc5abad3.js';

console.log(x);
console.log(\\"content.js\\");
chrome.contextMenus.removeAll();
"
`;

exports[`manifest vs output: devtools/devtools.html 1`] = `
"<!DOCTYPE html><html><head>
	<script src=\\"devtools2.js\\" type=\\"module\\"></script>
	<script src=\\"devtools1.js\\" type=\\"module\\"></script>
</head>
<body>
</body></html>"
`;

exports[`manifest vs output: devtools/devtools1.js 1`] = `
"const message = \\"devtools.ts\\";
console.log(message);

export { message };
"
`;

exports[`manifest vs output: devtools/devtools2.js 1`] = `
"import { x } from '../modules/imported-fc5abad3.js';

console.log(x);
console.log(\\"background.js\\");
chrome.storage.local.clear();
"
`;

exports[`manifest vs output: manifest.json 1`] = `"{\\"name\\":\\"mv2-kitchen-sink\\",\\"description\\":\\"mv2-kitchen-sink chrome extension\\",\\"manifest_version\\":2,\\"version\\":\\"1.0.0\\",\\"icons\\":{\\"16\\":\\"images/icon-main-16.png\\",\\"48\\":\\"images/icon-main-48.png\\",\\"128\\":\\"images/icon-main-128.png\\"},\\"default_locale\\":\\"en\\",\\"devtools_page\\":\\"devtools/devtools.html\\",\\"background\\":{\\"scripts\\":[\\"background.esm-wrapper.js\\"]},\\"permissions\\":[\\"webRequest\\",\\"webRequestBlocking\\",\\"storage\\",\\"contextMenus\\",\\"bookmarks\\",\\"cookies\\"],\\"content_scripts\\":[{\\"js\\":[\\"assets/content.esm-wrapper-a6ac6bbe.js\\"],\\"matches\\":[\\"https://www.google.com/*\\"]},{\\"js\\":[\\"content.js\\"],\\"css\\":[\\"content.css\\"],\\"matches\\":[\\"https://www.yahoo.com/*\\"]}],\\"options_page\\":\\"options.html\\",\\"browser_action\\":{\\"default_icon\\":{\\"16\\":\\"images/icon-main-16.png\\"},\\"default_popup\\":\\"popup/popup.html\\"},\\"web_accessible_resources\\":[\\"options.jpg\\",\\"fonts/*.ttf\\",\\"fonts/*.otf\\",\\"content.js\\"],\\"content_security_policy\\":\\"script-src 'self'; object-src 'self'\\"}"`;

exports[`manifest vs output: modules/imported-fc5abad3.js 1`] = `
"const x = \\"x\\";

export { x };
"
`;

exports[`manifest vs output: options.css 1`] = `
"body {
    color: red;
  }
  "
`;

exports[`manifest vs output: options.html 1`] = `
"<!DOCTYPE html><html lang=\\"en\\"><head>
    <meta charset=\\"UTF-8\\">
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\">
    <meta http-equiv=\\"X-UA-Compatible\\" content=\\"ie=edge\\">
    <title>Options</title>
    <link rel=\\"stylesheet\\" href=\\"options.css\\">
    <link rel=\\"icon\\" href=\\"images/favicon.png\\">
  </head>

  <body>
    <!-- Script chunks: to bundle -->
    <script src=\\"./options1.js\\" type=\\"module\\"></script>
    <script src=\\"./options2.js\\" type=\\"module\\"></script>
    <script src=\\"./options3.js\\" type=\\"module\\"></script>
    <script src=\\"./options4.js\\" type=\\"module\\"></script>
    <!-- Image assets -->
    <img src=\\"./options.png\\">
    <img src=\\"./options.jpg\\">
    <!-- Script asset: do not bundle -->
    <script src=\\"./asset.js\\" data-rollup-asset=\\"true\\"></script>
  

</body></html>"
`;

exports[`manifest vs output: options1.js 1`] = `
"console.log(\\"options1.js\\");
chrome.bookmarks.get(\\"main\\");
"
`;

exports[`manifest vs output: options2.js 1`] = `
"console.log(\\"options2.jsx\\");
chrome.cookies.getAll();
"
`;

exports[`manifest vs output: options3.js 1`] = `
"console.log(\\"options3.ts\\");
"
`;

exports[`manifest vs output: options4.js 1`] = `
"console.log(\\"options4.tsx\\");
"
`;

exports[`manifest vs output: popup/popup.html 1`] = `
"<!DOCTYPE html><html lang=\\"en\\"><head>
    <meta charset=\\"UTF-8\\">
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\">
    <meta http-equiv=\\"X-UA-Compatible\\" content=\\"ie=edge\\">
    <link rel=\\"icon\\" href=\\"/images/favicon.ico\\">
    <title>Popup</title>
  </head>

  <body>
    <script src=\\"popup.js\\" type=\\"module\\"></script>
  

</body></html>"
`;

exports[`manifest vs output: popup/popup.js 1`] = `
"console.log(\\"popup.js\\");
"
`;
